html/docs/design/sync-ot.html

```diff
@@ -0,0 +1,435 @@
+                </p>
+              </li>
+            </ul>
+            </p>
+
+            <h4 id="servers">Singleton Replicas</h4>
+
+            <p>
+            A replica depends on periodic synchronization for performant operation.
+            Without synchronization, its list of pending operations would grow indefinitely, and tasks could never be expired.
+            So all replicas, even "singleton" replicas which do not replicate task data with any other replica, must synchronize periodically.
+            To support this condition, Taskwarrior provides a local implementation of this model against which sync operations can be performed, and a migration path from such local sync to remote synchronization.
+            </p>
+
+            <a name="model"></a>
+            <h3>Implementation</h3>
+
+            <p>
+            Achieving this approach in Taskwarrior requires some difficult changes.
+            </p>
+
+            <ul>
+              <li>
+                <p>
+                All modifications to tasks must be made in the form of operations.
+                This is not conceptually difficult, but represents a significant refactor of the existing code.
+                The most challenging aspect of this change has <a href="https://github.com/GothenburgBitFactory/taskwarrior/issues/2553">already</a> <a href="https://github.com/GothenburgBitFactory/taskwarrior/issues/2569">been completed</a>: changes such as adding or removing a tag must be formulated as single operations that can be easily transposed.
+                For example, consider conflicting operations <code>Update($uuid, "tags", "foo,bar,bing")</code> and <code>Update($uuid, "tags", "foo,bar,next")</code>.
+                Without an understanding of tags, transposing these two operations requires one to "win", effectively undoing the other.
+                If, instead, each tag is stored as a distinct task key, then the two operations to add tags are <code>Update($uuid, "tag_bing", "")</code> and <code>Update($uuid, "tag_next", "")</code>.
+                These operations do not conflict with each other and can be transposed easily.
+                </p>
+              </li>
+
+              <li>
+                <p>
+                Synchronization becomes a required kind of periodic maintenance, even if against a local implementation.
+                This may be performed automatically, or existing users may be required to run <code>task sync</code> periodically.
+                </p>
+              </li>
+
+              <li>
+                <p>
+                The current on-disk storage for a replica is text-based and invites users to edit it.
+                This will interact particularly badly with operation-based synchronization, so it may be necessary to choose a more robust format.
```

Comment:
I'd claim this is orthogonal, in fact, the current data storage format has exactly the same amount of danger. The data about the tasks is stored in `pending.data`, the task modifications (akin to the operations proposed to in this document are stored in `backlog.data`). Modifying `pending.data` manually and expecting sync to work would lead to the same data corruption as described here even with the current implementation.

I went over (triaged / reproduced where applicable) several hundred issues in the past year. For what it's worth, so far the number of bugs that I've encountered whose root problem was hand-editing the `.data` files was 0. While I support the exploration of alternative data stores as outlined in the discussion [about atomic data file writes](https://github.com/GothenburgBitFactory/taskwarrior/issues/152) I would say the evidence that this is something that users do is not there.

Reply:
Ah, that's good data.  I'll remove this section.